### ISSUE TITLE
cargo lies about host name resolution.

### ISSUE THREAD

**Issue by smosher** [rust-lang/cargo#1192]
I frequently get this kind of problem:

    $ cargo build --verbose
    Downloading ao v0.3.0
    Unable to get packages from source

    Caused by:
      Failed to download package `ao v0.3.0` from https://crates.io/api/v1/crates/ao/0.3.0/download

    Caused by:
      Couldn't resolve host name

However, the host in the error does resolve (crates.io) and loading the URL in a browser gives me this:

    Oops, that route doesn't exist!

There's clearly an error, but not the one reported.

Aside: I'm not clear on why cargo is talking to crates.io (since it was told to fetch from bitbucket in this case), or under what circumstances it's meant to do that. But that's a separate concern.

Edit: 

    $ cargo --version
    cargo 0.0.1-pre-nightly (8154255 2015-01-14 19:14:12 +0000)

**Comment by alexcrichton** [comment]
This is interesting! We've had some DNS difficulties with crates.io in the past. Next time you encounter this error, can you run this command and paste the output here?

```
for letter in a b c d e f; do echo "$letter.ns.zerigo.net"; dig +noall +answer crates.io @$letter.ns.zerigo.net; echo; done
```

**Comment by smosher** [comment]
Sure. Is cargo doing name resolution differently? I believe I checked into this before with cargo being the only thing that complained of resolution issues on crates.io.

It *is* working at the moment. Is the url reported by cargo when there is an error meaningful at all? If I go there in a browser I still get the same route error. Up to now I thought whatever that meant was being misreported by cargo... though I did change the .toml since then.

**Comment by alexcrichton** [comment]
Cargo is currently using libcurl for HTTP requests and I believe that name resolution is happening in libcurl (we're certainly not doing anything special!).

The error you're seeing from that URL is indeed intended as it's meant to be requested with JSON rather than HTML (the server isn't super great with respect to this).

**Comment by smosher** [comment]
That's confusing¹ but ultimately this is good news. It means I should be able to test it end-to-end and see whether other tools can fetch the crate.

¹ It seems wget will fetch a .tar.gz, why not just deliver that to the browser and let the user decline the download if that's not what they want? I had also been confused about the lack of download links on crates.io... seems that wouldn't work anyway because of this behaviour.

**Comment by alexcrichton** [comment]
Yeah I suspect the case should be handled by the frontend a bit better. Currently we're not providing download links as there's not really much you can do with a `.crate` file (we don't provide the tools to manage it)

**Comment by smosher** [comment]
So I just had a weird DNS issue that seemed to only affect crates.io. The problem was local to the caching DNS server, querying other name servers worked fine. But I noticed the TTL seems to be set to 60. Is that a good idea?

This could explain the problem with some tools (eg. browsers) being able to resolve the site and others not. If so, it looks like I was too hasty with the title. Until I saw that lowball TTL I just couldn't believe this was a DNS issue.
